python-stdlib/logging/logging.py

```diff
@@ -1,128 +1,202 @@
 import sys
+import time
 
-CRITICAL = 50
-ERROR = 40
-WARNING = 30
-INFO = 20
-DEBUG = 10
 NOTSET = 0
+DEBUG = 10
+INFO = 20
+WARNING = 30
+ERROR = 40
+CRITICAL = 50
 
-_level_dict = {
-    CRITICAL: "CRIT",
-    ERROR: "ERROR",
-    WARNING: "WARN",
-    INFO: "INFO",
+levelname = {
+    NOTSET: "NOTSET",
     DEBUG: "DEBUG",
+    INFO: "INFO",
+    WARNING: "WARNING",
+    ERROR: "ERROR",
+    CRITICAL: "CRITICAL",
 }
 
-_stream = sys.stderr
+loggers = {}
+default_fmt = "%(levelname)s:%(name)s:%(message)s"
+default_datefmt = "%Y-%m-%d %H:%M:%S"
 
 
-class LogRecord:
-    def __init__(self):
-        self.__dict__ = {}
+class Handler:
+    def __init__(self, level=NOTSET):
+        self.level = level
+        self.formatter = None
 
-    def __getattr__(self, key):
-        return self.__dict__[key]
+    def close(self):
+        pass
 
+    def setLevel(self, level):
+        self.level = level
 
-class Handler:
-    def __init__(self):
-        pass
+    def setFormatter(self, formatter):
+        self.formatter = formatter
 
-    def setFormatter(self, fmtr):
-        pass
+    def format(self, record):
+        return self.formatter.format(record)
 
 
-class Logger:
+class StreamHandler(Handler):
+    def __init__(self, stream=sys.stderr):
+        self.stream = stream
+        self.terminator = "\n"
 
-    level = NOTSET
-    handlers = []
-    record = LogRecord()
+    def close(self):
+        if hasattr(self.stream, "flush"):
+            self.stream.flush()
 
-    def __init__(self, name):
+    def emit(self, record):
+        if record.level >= self.level:
+            self.stream.write(self.format(record) + self.terminator)
+
+
+class FileHandler(StreamHandler):
+    def __init__(self, filename, mode="a", encoding="UTF-8"):
+        super().__init__(stream=open(filename, mode=mode, encoding=encoding))
+
+    def close(self):
+        super().close()
+        self.stream.close()
+
+
+class Record:
+    def set(self, name, level, message):
         self.name = name
+        self.level = level
+        self.message = message
+        self.ts = time.time()
+        self.msecs = int((self.ts - int(self.ts)) * 1000)
+
+
+class Formatter:
+    def __init__(self, fmt=default_fmt, datefmt=default_datefmt):
+        self.fmt = fmt
+        self.datefmt = datefmt
+
+    def formatTime(self, datefmt):
+        if hasattr(time, "strftime"):
+            return time.strftime(datefmt, time.localtime())
+        return ""
+
+    def format(self, record):
+        return self.fmt % {
+            "name": record.name,
+            "message": record.message,
+            "msecs": record.msecs,
+            "asctime": self.formatTime(self.datefmt),
+            "levelname": levelname[record.level],
+        }
```

Comment:
In the original code, the reason to have `__dict__` was so this could be written as: `self.fmt % record.__dict__`.

Reply:
Well, we could `return self.fmt % record.__dict__` ?

Reply:
Yes it could be changed to that, that would work, but remember that calling `.__dict__` creates a read-only copy every time!  The original code had an explicit `__dict__` created so that accessing this member did not need to make a copy.

Reply:
Okay so keeping it as it is for now.